File: lib/ios/VesperPlayerKit/Sources/VesperPlayerKit/Playback/VesperHdrOutput.swift
import Combine
import Foundation

/// Observed display output. Capability and decoder metadata cannot confirm this state.
public enum VesperHdrOutputState: String, Sendable {
    case unknown, sdr, hdr
}

public enum VesperHdrOutputFormat: String, Sendable {
    case unknown, hdr10, hlg, dolbyVision
}

/// Output evidence and its native lifecycle context. Revisions are local to one
/// controller; sourceRevision counts activations, including repeated URLs.
/// iOS currently reports unknown because no per-player display observer is enabled.
public struct VesperHdrOutputSnapshot: Equatable, Sendable {
    public internal(set) var state: VesperHdrOutputState = .unknown
    public internal(set) var format: VesperHdrOutputFormat = .unknown
    public internal(set) var sourceRevision: Int64 = 0
    public internal(set) var outputGeneration: Int64 = 0
    public internal(set) var effectiveVideoTrackId: String?
    public internal(set) var catalogRevision: Int64?
    public internal(set) var displayId: String?
    public internal(set) var evidence: String?
    public internal(set) var reason: String? = "outputObservationUnavailable"
}

struct VesperHdrOutputObservationToken {
    fileprivate let owner: UUID
    let sourceRevision: Int64
    let outputGeneration: Int64
}

struct VesperHdrOutputObservation {
    let state: VesperHdrOutputState
    var format: VesperHdrOutputFormat = .unknown
    var evidence: String?
    var reason: String?
}

@MainActor
final class VesperHdrOutputTracker {
    private let owner = UUID()
    private var disposed = false
    @Published private(set) var snapshot = VesperHdrOutputSnapshot()

    func capture() -> VesperHdrOutputObservationToken {
        VesperHdrOutputObservationToken(
            owner: owner,
            sourceRevision: snapshot.sourceRevision,
            outputGeneration: snapshot.outputGeneration
        )
    }

    func sourceChanged() {
        guard !disposed else { return }
        var context = snapshot
        context.sourceRevision += 1
        context.effectiveVideoTrackId = nil
        context.catalogRevision = nil
        invalidate(context)
    }

    func videoTrackChanged(_ trackId: String?, catalogRevision: Int64?) {
        guard snapshot.effectiveVideoTrackId != trackId || snapshot.catalogRevision != catalogRevision else { return }
        var context = snapshot
        context.effectiveVideoTrackId = trackId
        context.catalogRevision = catalogRevision
        invalidate(context)
    }

    func outputPathChanged() {
        invalidate(snapshot)
    }

    private func invalidate(_ context: VesperHdrOutputSnapshot) {
        guard !disposed else { return }
        var next = context
        next.outputGeneration = snapshot.outputGeneration + 1
        next.state = .unknown
        next.format = .unknown
        next.evidence = nil
        next.reason = "outputObservationUnavailable"
        snapshot = next
    }

    @discardableResult
    func apply(_ token: VesperHdrOutputObservationToken, observation: VesperHdrOutputObservation) -> Bool {
        guard !disposed, token.owner == owner,
              token.sourceRevision == snapshot.sourceRevision,
              token.outputGeneration == snapshot.outputGeneration else { return false }
        guard observation.state == .unknown || observation.evidence?.isEmpty == false else { return false }
        var next = snapshot
        next.state = observation.state
        next.format = observation.state == .hdr ? observation.format : .unknown
        next.evidence = observation.state == .unknown ? nil : observation.evidence
        next.reason = observation.reason
        snapshot = next
        return true
    }

    func dispose() {
        guard !disposed else { return }
        outputPathChanged()
        disposed = true
    }
}

File: lib/android/vesper-player-kit/src/main/java/io/github/umbrella22/vesper/player/android/VesperHdrOutput.kt
package io.github.umbrella22.vesper.player.android

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** Observed display output. Capability and decoder metadata cannot confirm this state. */
enum class VesperHdrOutputState(val wireName: String) {
    Unknown("unknown"), Sdr("sdr"), Hdr("hdr"),
}

enum class VesperHdrOutputFormat(val wireName: String) {
    Unknown("unknown"), Hdr10("hdr10"), Hlg("hlg"), DolbyVision("dolbyVision"),
}

/**
 * Output evidence and its native lifecycle context. Revisions are local to one
 * controller; sourceRevision counts source activations, including repeated URLs.
 * Android currently reports unknown because no per-Surface output observer is enabled.
 */
@ConsistentCopyVisibility
data class VesperHdrOutputSnapshot internal constructor(
    val state: VesperHdrOutputState = VesperHdrOutputState.Unknown,
    val format: VesperHdrOutputFormat = VesperHdrOutputFormat.Unknown,
    val sourceRevision: Long = 0,
    val outputGeneration: Long = 0,
    val effectiveVideoTrackId: String? = null,
    val catalogRevision: Long? = null,
    val displayId: String? = null,
    val evidence: String? = null,
    val reason: String? = "outputObservationUnavailable",
)

internal data class VesperHdrOutputObservationToken(
    val owner: Any,
    val sourceRevision: Long,
    val outputGeneration: Long,
)

internal data class VesperHdrOutputObservation(
    val state: VesperHdrOutputState,
    val format: VesperHdrOutputFormat = VesperHdrOutputFormat.Unknown,
    val evidence: String? = null,
    val reason: String? = null,
)

/** Mutated on the player looper. Tokens belong to this instance, never to recurring IDs. */
internal class VesperHdrOutputTracker(hasInitialSource: Boolean = false) {
    private val owner = Any()
    private var disposed = false
    private var outputListener: ((VesperHdrOutputSnapshot) -> Unit)? = null
    private val mutableSnapshot = MutableStateFlow(
        VesperHdrOutputSnapshot(
            sourceRevision = if (hasInitialSource) 1 else 0,
            outputGeneration = if (hasInitialSource) 1 else 0,
        ),
    )
    val snapshot: StateFlow<VesperHdrOutputSnapshot> = mutableSnapshot.asStateFlow()

    fun setListener(listener: ((VesperHdrOutputSnapshot) -> Unit)?) {
        if (disposed) return
        outputListener = listener
        listener?.invoke(snapshot.value)
    }

    private fun publish(next: VesperHdrOutputSnapshot) {
        if (next == snapshot.value) return
        mutableSnapshot.value = next
        // StateFlow may conflate transitions; the boundary listener must see each one.
        outputListener?.invoke(next)
    }

    fun capture(): VesperHdrOutputObservationToken = snapshot.value.let {
        VesperHdrOutputObservationToken(owner, it.sourceRevision, it.outputGeneration)
    }

    fun sourceChanged() {
        if (disposed) return
        invalidate(snapshot.value.copy(
            sourceRevision = snapshot.value.sourceRevision + 1,
            effectiveVideoTrackId = null,
            catalogRevision = null,
        ))
    }

    fun videoTrackChanged(trackId: String?, catalogRevision: Long?) {
        val current = snapshot.value
        if (current.effectiveVideoTrackId == trackId && current.catalogRevision == catalogRevision) return
        invalidate(current.copy(effectiveVideoTrackId = trackId, catalogRevision = catalogRevision))
    }

    fun outputPathChanged(displayId: String? = snapshot.value.displayId) {
        invalidate(snapshot.value.copy(displayId = displayId))
    }

    private fun invalidate(context: VesperHdrOutputSnapshot) {
        if (disposed) return
        publish(context.copy(
            outputGeneration = snapshot.value.outputGeneration + 1,
            state = VesperHdrOutputState.Unknown,
            format = VesperHdrOutputFormat.Unknown,
            evidence = null,
            reason = "outputObservationUnavailable",
        ))
    }

    fun apply(token: VesperHdrOutputObservationToken, observation: VesperHdrOutputObservation): Boolean {
        val current = snapshot.value
        if (disposed || token.owner !== owner || token.sourceRevision != current.sourceRevision ||
            token.outputGeneration != current.outputGeneration
        ) return false
        if (observation.state != VesperHdrOutputState.Unknown && observation.evidence.isNullOrBlank()) return false
        publish(current.copy(
            state = observation.state,
            format = if (observation.state == VesperHdrOutputState.Hdr) observation.format else VesperHdrOutputFormat.Unknown,
            evidence = observation.evidence.takeIf { observation.state != VesperHdrOutputState.Unknown },
            reason = observation.reason,
        ))
        return true
    }

    fun dispose() {
        if (disposed) return
        outputPathChanged(displayId = null)
        disposed = true
        outputListener = null
    }
}
